pkg/porter/uninstall.go

```diff
@@ -32,16 +40,41 @@ func (p *Porter) UninstallBundle(opts UninstallOptions) error {
 		return err
 	}
 
+	var uninstallErrs error
 	fmt.Fprintf(p.Out, "uninstalling %s...\n", opts.Name)
 	err = p.CNAB.Execute(opts.ToActionArgs(deperator))
 	if err != nil {
+		uninstallErrs = multierror.Append(uninstallErrs, err)
+		if strings.Contains(err.Error(), claim.ErrInstallationNotFound.Error()) {
+			return uninstallErrs
+		}
+
 		if len(deperator.deps) > 0 {
-			return errors.Wrapf(err, "failed to uninstall the %s bundle, the remaining dependencies were not uninstalled", opts.Name)
-		} else {
-			return err
+			uninstallErrs = multierror.Append(uninstallErrs,
+				fmt.Errorf("failed to uninstall the %s bundle, the remaining dependencies were not uninstalled", opts.Name))
+		}
+
+		if opts.Delete && !opts.ForceDelete {
+			uninstallErrs = multierror.Append(uninstallErrs, fmt.Errorf("not deleting installation %s as uninstall was not successful; use --force-delete to override", opts.Name))
 		}
+		if !opts.ForceDelete {
+			return uninstallErrs
+		}
+		// else, we swallow uninstallErrs as opts.ForceDelete is true
+		// and we wish to pass same option down to deps, if applicable
 	}
 
 	// TODO: See https://github.com/deislabs/porter/issues/465 for flag to allow keeping around the dependencies
-	return deperator.Execute()
+	err = deperator.Execute()
+	if err != nil {
```

Comment:
Right now it's confusing how the ForceDelete/Delete is handled by both this function and the Deperator, and then they have all sorts of conditionals for handling the error, printing that warning about using --force, etc. I am concerned that it's going to be difficult to maintain without introducing bugs or fixing a bug accidentally in just one place (and forgetting to fix it in the dependencies.go for example).

🤔 Looking at it in a github code review it's hard to say "just change X" to simplify it. But I do wish there was a way to do so. Maybe we can take a second pass at it and see if there's a way to either extract the logic for dealing with it, or consolidate the logic for checking and printing the messages? I'm not sure what to suggest which is why I think trying again in a follow-up or whenever we next feel clever or inspired is fine.

One idea is to have a function takes in the flags (delete, force delete and the installation status) and it handles deciding if we should delete the installation and printing the warning. Then at least we could reuse that in 2+ places?

Reply:
You're right, this area warrants further thought.  Thank you for the initial suggestions.  I'll spend some more time here and hopefully have an update w/ less fragility and duplication.

Reply:
(Still working on this one...)

Reply:
Ok, refactored for a bit of simplification... or at least some shared logic.  Not sure if better, but perhaps it might generate some further ideas: https://github.com/deislabs/porter/pull/1189/commits/dd5dd04dba74654b626a5ccf92eec5ebcc52b1b3